consensus/XDPoS/engines/engine_v2/engine.go

```diff
@@ -1006,6 +1021,20 @@ func (x *XDPoS_v2) calcMasternodes(chain consensus.ChainReader, blockNum *big.In
 		return nil, nil, err
 	}
 	masternodes := common.RemoveItemFromArray(candidates, penalties)
+	if len(masternodes) > maxMasternodes {
+		masternodes = masternodes[:maxMasternodes]
+	}
+	if len(masternodes) < x.config.V2.CurrentConfig.CertThreshold {
+		log.Warn("[calcMasternodes] Next epoch masternodes less than threshold", "number", blockNum, "masternodes", len(masternodes), "threshold", x.config.V2.CurrentConfig.CertThreshold)
+		for i, a := range masternodes {
+			log.Warn("final masternode", "i", i, "addr", a)
+		}
+		if len(penalties) > 0 {
```

Comment:
we can delete this line, if penalties is empty, then for loop would stop immediately

Reply:
fixed